To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

In [ ]:
completions = [
    [{"role": "assistant", "content": "<think>The sum of 1 and 2 is 3, which we multiply by 4 to get 12.</think><answer>(1 + 2) * 4 = 12</answer>"},{"role": "assistant", "content": "<think>The sum of 1 and 2 is 3, which we multiply by 4 to get 12.</think><answer>(1 + 2) * 4 = 12</answer>"} ],
    [{"role": "assistant", "content": "The sum of 3 and 1 is 4, which we multiply by 2 to get 8. So (3 + 1) * 2 = 8."}],
]
responses = [completion[0]["content"] for completion in completions]
print(responses)

['<think>The sum of 1 and 2 is 3, which we multiply by 4 to get 12.</think><answer>(1 + 2) * 4 = 12</answer>', 'The sum of 3 and 1 is 4, which we multiply by 2 to get 8. So (3 + 1) * 2 = 8.']


In [ ]:
prompts = [
    [{"role": "assistant", "content": "What is the result of (1 + 2) * 4?"}],
    [{"role": "assistant", "content": "What is the result of (3 + 1) * 2?"}],
]

question = prompts[0][-1]["content"]
print(question)

What is the result of (1 + 2) * 4?


**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [ ]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    !pip install --no-deps unsloth vllm

In [ ]:
#@title Colab Extra Install { display-mode: "form" }
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    !pip install --no-deps unsloth vllm
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    # Skip restarting message in Colab
    import sys, re, requests; modules = list(sys.modules.keys())
    for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft "trl==0.15.2" triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer

    # vLLM requirements - vLLM breaks Colab due to reinstalling numpy
    f = requests.get("https://raw.githubusercontent.com/vllm-project/vllm/refs/heads/main/requirements/common.txt").content
    with open("vllm_requirements.txt", "wb") as file:
        file.write(re.sub(rb"(transformers|numpy|xformers)[^\n]{1,}\n", b"", f))
    !pip install -r vllm_requirements.txt

### Unsloth

Load up `Llama 3.1 8B Instruct`, and set parameters

In [ ]:
from unsloth import FastLanguageModel
import torch
import json
from transformers import AutoTokenizer, AutoModelForCausalLM
from trl import GRPOTrainer, GRPOConfig
from datasets import load_dataset

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 05-04 04:29:35 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-04 04:29:35 [__init__.py:239] Automatically detected platform cuda.


In [ ]:
import os, wandb
from huggingface_hub import login

os.environ["WANDB_API_KEY"] = "810a489e4029d94651a3868c08a49812cc1e4cc7"
os.environ["HUGGINGFACE_API_KEY"] = "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"

In [ ]:
hf_token = os.environ["HUGGINGFACE_API_KEY"]
login(hf_token)

secret_key = os.environ["WANDB_API_KEY"]
wandb.login(key=secret_key)

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: sonhh2003 (sonhh2003-hanoi-university-of-science-and-technology) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

### Load Tokenizer & Model

In [ ]:
max_seq_length = 1024 # Can increase for longer reasoning traces
lora_rank = 16 # Larger rank = smarter, but slower
model_name = "meta-llama/meta-Llama-3.1-8B-Instruct"    # Gemma 3.1 8B‑Instruct checkpoint
output_dir = "./grpo_llama3_8b"

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = model_name,
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.6, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        # "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ], # Remove QKVO if out of memory
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

==((====))==  Unsloth 2025.4.7: Fast Llama patching. Transformers: 4.51.3. vLLM: 0.8.5.post1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/meta-llama-3.1-8b-instruct-unsloth-bnb-4bit with actual GPU utilization = 59.43%
Unsloth: Your GPU has CUDA compute capability 7.5 with VRAM = 14.74 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 1024. Num Sequences = 160.
Unsloth: vLLM's KV Cache can use up to 2.66 GB. Also swap space = 2 GB.
WARNING 05-04 04:29:47 [config.py:2972] Casting torch.bfloat16 to torch.float16.
INFO 05-04 04:30:16 [config.py:717] This model supports multiple tasks: {'classify', 'generate', '

tokenizer_config.json:   0%|          | 0.00/55.5k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/454 [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

INFO 05-04 04:30:30 [cuda.py:240] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
INFO 05-04 04:30:30 [cuda.py:289] Using XFormers backend.
INFO 05-04 04:30:31 [parallel_state.py:1004] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, TP rank 0
INFO 05-04 04:30:31 [model_runner.py:1108] Starting to load model unsloth/meta-llama-3.1-8b-instruct-unsloth-bnb-4bit...
INFO 05-04 04:30:31 [loader.py:1187] Loading weights with BitsAndBytes quantization. May take a while ...
INFO 05-04 04:30:32 [weight_utils.py:265] Using model weights format ['*.safetensors']


model.safetensors:   0%|          | 0.00/5.96G [00:00<?, ?B/s]

INFO 05-04 04:34:37 [weight_utils.py:281] Time spent downloading weights for unsloth/meta-llama-3.1-8b-instruct-unsloth-bnb-4bit: 244.812036 seconds
INFO 05-04 04:34:37 [weight_utils.py:315] No model.safetensors.index.json found in remote.


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 05-04 04:35:26 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 05-04 04:35:27 [model_runner.py:1140] Model loading took 5.6877 GiB and 295.108242 seconds
INFO 05-04 04:35:38 [worker.py:287] Memory profiling takes 10.94 seconds
INFO 05-04 04:35:38 [worker.py:287] the current vLLM instance can use total_gpu_memory (14.74GiB) x gpu_memory_utilization (0.59) = 8.76GiB
INFO 05-04 04:35:38 [worker.py:287] model weights take 5.69GiB; non_torch_memory takes 0.03GiB; PyTorch activation peak memory takes 0.74GiB; the rest of the memory reserved for KV Cache is 2.30GiB.
INFO 05-04 04:35:39 [executor_base.py:112] # cuda blocks: 1178, # CPU blocks: 1024
INFO 05-04 04:35:39 [executor_base.py:117] Maximum concurrency for 1024 tokens per request: 18.41x
INFO 05-04 04:35:40 [model_runner.py:1450] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI. I

Capturing CUDA graph shapes:   0%|          | 0/23 [00:00<?, ?it/s]

INFO 05-04 04:36:40 [model_runner.py:1592] Graph capturing finished in 60 secs, took 0.53 GiB
INFO 05-04 04:36:40 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 73.42 seconds


tokenizer_config.json:   0%|          | 0.00/55.5k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/454 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

Not an error, but Unsloth cannot patch Attention layers with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Not an error, but Unsloth cannot patch O projection layer with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Unsloth 2025.4.7 patched 32 layers with 0 QKV layers, 0 O layers and 32 MLP layers.


In [ ]:
print(tokenizer.model_max_length)

131072


### Data Prep
<a name="Data"></a>

We directly leverage [@willccbb](https://gist.github.com/willccbb/4676755236bb08cab5f4e54a0475d6fb) for data prep and all reward functions. You are free to create your own!

In [ ]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.4/491.4 kB 9.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 15.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 16.1 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.2
    Uninstalling fsspec-2025.3.2:
      Successfully uninstalled fsspec-2025.3.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.2 requires fsspec==2025.3.2, but you have fsspec 2025.3.0 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which i

In [ ]:
from datasets import Dataset, load_dataset

SYSTEM_PROMPT = """
ROLE: You are a cookie policy analysis expert.
Your task is to read a cookie policy text and extract detailed information about each cookie mentioned in that policy.
RESPONSE Format: If no cookies are specifically described, return []. Otherwise, return the result as a JSON array following structure:
  [
    {
      "cookie_name": "cookie_name",
      "declared_purpose": "declared_purpose",
      "declared_retention": "declared
      "declared_third_parties": ["declared_third_parties"],
      "declared_description": "declared_description"
    }
  ]

Specific Requirements:
  Read and Analyze the Text: Carefully read the entire content of the cookie policy to understand the types of cookies used, their purposes, storage duration, detailed descriptions, and information about the owner (first party or third party).
  Identify Described Cookies: Search for sections, paragraphs, or text that describe specific cookies. Each cookie will usually have its own entry or be listed in a list.

  Extract Detailed Information for Each Cookie: For each identified cookie, carefully extract the following information from the policy:
    +) "cookie_name": Find the technical name or descriptive name of the cookie exactly as it is mentioned in the policy.
    +) "declared_purpose": A detailed description of the cookie's purpose. With the following options:
        +) Strictly Necessary: Cookies essential for the website to function correctly with basic features (e.g., managing login sessions, shopping cart functionality, secure navigation).
        +) Functionality: Cookies that help personalize the user experience and provide additional features (e.g., remembering language preferences, page layouts).
        +) Analytical: Cookies used to collect data and understand user behavior on the website (e.g., number of visits, most viewed pages).
        +) Targeting/Advertising/Marketing: Cookies used to display personalized advertisements based on user behavior or interests.
        +) Performance: Cookies that help improve the technical performance of the website (e.g., optimizing page loading, content distribution).
        +) Social Sharing: Cookies that enable sharing content from the website to social media platforms.
        +) Null if the policy does not provide any specific information about the purpose of this cookie.
    +) "declared_retention": Find information about the duration this cookie will be stored on the user's device (e.g., "6 months", "24 hours", "Session", "Persistent", "1 minute", "Until deleted"). Record this information as accurately as possible according to the original text.
    +) "declared_third_parties": An array of third parties involved in the use of this cookie (if any). ["First Party"] indicates a first-party cookie set by the website itself.
    +) "declared_description": The content taken directly from the text, without fabrication, using exactly what they have described.
"""

def preprocess(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    prompt = [
        {'role': 'system', 'content': SYSTEM_PROMPT},
        {'role': 'user', 'content': content}
    ]

    return {
        'prompt': prompt,
        'answer': example['label']
        # 'answer': '[]'
    }

dataset = load_dataset("sonhask/detect-cookie-with-label")
dataset = dataset['train'].map(preprocess)


Map:   0%|          | 0/7143 [00:00<?, ? examples/s]

In [ ]:
def compute_token_length(example):
    length =  (
              len(example['answer']) +
               len(example["prompt"][0]['content']) + len(example["prompt"][1]['content']))*0.331 + 50
    return length

dataset = dataset.filter(lambda example: compute_token_length(example) <= 2048)
len(dataset)

Filter:   0%|          | 0/7143 [00:00<?, ? examples/s]

4136

In [ ]:
def thong_ke(dataset):
    empty_label_count = 0
    non_empty_label_count = 0

    for i in dataset:
        if i['answer'] == '[]':
            empty_label_count += 1
        else:
            non_empty_label_count += 1

    print("#[]: ", empty_label_count)
    print("#Has cookies: ", non_empty_label_count)

thong_ke(dataset)

#[]:  3817
#Has cookies:  319


In [ ]:
# 2. remove every column except the one GRPOTrainer expects:
keep = ["prompt", "answer"]
to_remove = [c for c in dataset.column_names if c not in keep]
dataset = dataset.remove_columns(to_remove)

# 3. now dataset.features == ['prompt']
print(dataset.features)

{'prompt': [{'content': Value(dtype='string', id=None), 'role': Value(dtype='string', id=None)}], 'answer': Value(dtype='string', id=None)}


In [ ]:
dataset[0]

{'prompt': [{'content': '\nROLE: You are a cookie policy analysis expert. \nYour task is to read a cookie policy text and extract detailed information about each cookie mentioned in that policy.\nRESPONSE Format: If no cookies are specifically described, return []. Otherwise, return the result as a JSON array following structure:\n  [\n    {\n      "cookie_name": "cookie_name",\n      "declared_purpose": "declared_purpose",\n      "declared_retention": "declared\n      "declared_third_parties": ["declared_third_parties"],\n      "declared_description": "declared_description"\n    }\n  ]\n\nSpecific Requirements:\n  Read and Analyze the Text: Carefully read the entire content of the cookie policy to understand the types of cookies used, their purposes, storage duration, detailed descriptions, and information about the owner (first party or third party).\n  Identify Described Cookies: Search for sections, paragraphs, or text that describe specific cookies. Each cookie will usually have i

### Define reward functions

In [ ]:
import jsonschema
import json
from jsonschema import validate, ValidationError
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
from itertools import zip_longest

embedder = SentenceTransformer('all-MiniLM-L6-v2')  # small, fast

schema = {
    "type": "array",
    "items": {
        "type": "object",
        "properties": {
            "cookie_name": {"type": "string"},
            "declared_purpose": {"type": ["string", "null"]},
            "declared_retention": {"type": ["string", "null"]},
            "declared_third_parties": {"type": "array"},
            "declared_description": {"type": ["string", "null"]}
        },
        "required": ["cookie_name"]
    }
}

def strict_format_reward_func(completions, **kwargs):
    rewards = []
    for completion in completions:
        try:
            response = json.loads(completion[0]['content'])
            validate(instance=response, schema=schema)
            rewards.append(1.0)
        except (json.JSONDecodeError, ValidationError):
            rewards.append(-1.0)
    return rewards

# Hàm này kiểm tra xem đầu ra có thể được phân tích cú pháp thành JSON hợp lệ hay không, mà không cần xác thực theo schema.
def soft_format_reward_func(completions, **kwargs):
    rewards = []
    for completion in completions:
        try:
            json.loads(completion[0]['content'])
            rewards.append(0.5)
        except json.JSONDecodeError:
            rewards.append(-0.5)
    return rewards

# def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
#     rewards = []
#     for completion in completions:
#         try:
#             response = json.loads(completion[0]['content'])
#             validate(instance=response, schema=schema)
#             correct_items = sum(1 for r_item in response if r_item in answer)
#             total_items = len(answer)
#             score = correct_items / total_items if total_items > 0 else 0.0
#             rewards.append(score)
#         except (json.JSONDecodeError, ValidationError):
#             rewards.append(-1.0)
#     return rewards

def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
    rewards = []
    print("Answer is: ", answer)
    for completion, answer in zip(completions, answer):
        try:
            response = json.loads(completion[0]['content'])
            grouth_truth = json.loads(answer)
            validate(instance=response, schema=schema)
            correct_items = sum(1 for r_item in response if r_item in grouth_truth)
            total_items = len(grouth_truth)
            score = correct_items / total_items if total_items > 0 else 0.0
            rewards.append(score)
        except (json.JSONDecodeError, ValidationError):
            rewards.append(-1.0)
    return rewards

# Hàm này đánh giá mức độ đầy đủ của các trường trong mỗi mục của đầu ra, tính điểm dựa trên tỷ lệ các trường không rỗng.
def field_completeness_reward_func(completions, **kwargs) -> list[float]:
    rewards = []
    for comp in completions:
        try:
            arr = json.loads(comp[0]["content"])
            total = len(arr) * 5
            filled = 0
            for item in arr:
                for v in [
                    item.get("cookie_name"),
                    item.get("declared_purpose"),
                    item.get("declared_retention"),
                    item.get("declared_third_parties"),
                    item.get("declared_description"),
                ]:
                    if v not in (None, "", []):
                        filled += 1
            rewards.append(filled / total if total else 0.0)
        except Exception:
            rewards.append(-1.0)
    return rewards

# FIELD-LEVEL REWARD
def field_level_reward(prompts, completions, answer, **kwargs) -> list[float]:
    required_fields = [
        "cookie_name",
        "declared_purpose",
        "declared_retention",
        "declared_third_parties",
        "declared_description",
    ]
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]
    print('*'*20, f"Question:\n{question}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}")

    rewards = []
    for comp, answer in zip(completions, answer):
        print("Len of comp: ", len(comp))
        print("Answer: ", answer)
        print("---> No.0 ", comp[0]["content"])
        try:
            pred = json.loads(comp[0]["content"])
            grouth_truth = json.loads(answer)
        except Exception:
            rewards.append(-1.0)
            continue

        if not isinstance(pred, list):
            rewards.append(0.0)
            continue

        score = 0.0
        matched = 0

        for p_item, g_item in zip_longest(pred, grouth_truth, fillvalue={}): # Trường hợp p có số lượng phần tử khác với g_item thì sao?
            match = 0
            for f in required_fields:
                if p_item.get(f, None) == g_item.get(f):
                    match += 1
            score += match / len(required_fields)
            matched += 1

        rewards.append(score / matched if matched else 0.0)
    return rewards

# JACCARD ON THIRD PARTIES
def jaccard_third_party_reward(prompts, completions, answer, **kwargs) -> list[float]:
    rewards = []

    for comp, answer in zip(completions, answer):
        try:
            pred = json.loads(comp[0]["content"])
            grouth_truth = json.loads(answer)
        except Exception:
            rewards.append(-1.0)
            continue

        if not isinstance(pred, list):
            rewards.append(-1.0)
            continue

        scores = []
        for p_item, g_item in zip_longest(pred, grouth_truth, fillvalue={}):
            set_p = set(p_item.get("declared_third_parties", []))
            set_g = set(g_item.get("declared_third_parties", []))
            inter = len(set_p & set_g)
            union = len(set_p | set_g)
            scores.append(inter / union if union > 0 else 0.0)

        rewards.append(sum(scores) / len(scores) if scores else 0.0)
    return rewards

# def semantic_description_reward(prompts, completions, answer, **kwargs) -> list[float]:
#     rewards = []

#     for comp in completions:
#         try:
#             pred = json.loads(comp[0]["content"])
#         except:
#             rewards.append(0.0)
#             continue

#         sims = []
#         grouth_truth = json.loads(gold)
#         for p_item, g_item in zip_longest(pred, grouth_truth, fillvalue={}): # Trường hợp p_item có số lượng phần tử khác với g_item thì sao?
#             desc_p = p_item.get("declared_description", "")
#             desc_g = g_item.get("declared_description", "")
#             emb = embedder.encode([desc_p, desc_g])
#             sims.append(float(cosine_similarity([emb[0]], [emb[1]])[0][0]))
#         rewards.append(sum(sims) / len(sims) if sims else 0.0)
#     return rewards

def semantic_description_reward(prompts, completions, answer, **kwargs) -> list[float]:
    rewards = []
    for comp, answer in zip(completions, answer):
        try:
            pred = json.loads(comp[0]["content"])
            grouth_truth = json.loads(answer)
        except:
            rewards.append(-1.0)
            continue

        sims = []
        for p_item, g_item in zip(pred, grouth_truth):
            desc_p = p_item.get("declared_description", "")
            desc_g = g_item.get("declared_description", "")
            emb = embedder.encode([desc_p, desc_g])
            sims.append(float(cosine_similarity([emb[0]], [emb[1]])[0][0]))
        rewards.append(sum(sims) / len(sims) if sims else 0.0)
    return rewards

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [ ]:
run = wandb.init(
    project="Fine-tune Llama 3 8B Instruct GRPO",
)

In [ ]:
max_prompt_length = 256

from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "paged_adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 2,
    gradient_accumulation_steps = 4, # Increase to 4 for smoother training
    num_generations = 2, # Decrease if out of memory
    # max_prompt_length = max_prompt_length,
    # max_completion_length = max_seq_length - max_prompt_length,
    num_train_epochs = 1, # Set to 1 for a full training run
    # max_steps = 250,
    save_steps = 250,
    max_grad_norm = 0.1,
    report_to = "wandb", # Can use Weights & Biases
    output_dir = "llama_3_1_8B_GRPO",
    max_prompt_length=2048
)

And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [ ]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        strict_format_reward_func,
        soft_format_reward_func,
        correctness_reward_func,
        field_completeness_reward_func,
        field_level_reward,
        jaccard_third_party_reward,
        semantic_description_reward,
    ],
    args = training_args,
    train_dataset = dataset,
)

trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 5,886 | Num Epochs = 1 | Total steps = 1,471
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 28,311,552/8,000,000,000 (0.35% trained)


Answer is:  ['[]', '[]']
******************** Question:
Cookie policy: []
Table: [] 
Answer:
[] 
Response:
Unfortunately, it seems like you didn't provide the cookie policy text and table. Please provide both, and I'll do my best to extract the detailed information about each cookie according to the format you provided.
Len of comp:  1
Answer:  []
---> No.0  Unfortunately, it seems like you didn't provide the cookie policy text and table. Please provide both, and I'll do my best to extract the detailed information about each cookie according to the format you provided.
Len of comp:  1
Answer:  []
---> No.0  []
Answer is:  ['[]', '[]']
******************** Question:
Cookie policy: []
Table: [] 
Answer:
[] 
Response:
The provided text is empty; therefore, there are no cookies to describe.
Len of comp:  1
Answer:  []
---> No.0  The provided text is empty; therefore, there are no cookies to describe.
Len of comp:  1
Answer:  []
---> No.0  Unfortunately, it appears the provided text does no

Answer is:  ['[]', '[]']
******************** Question:
Cookie policy: []
Table: [] 
Answer:
[] 
Response:
Since the cookie policy text is empty, I have no cookies to analyze.
Len of comp:  1
Answer:  []
---> No.0  Since the cookie policy text is empty, I have no cookies to analyze.
Len of comp:  1
Answer:  []
---> No.0  It seems that the text you provided is empty.
Answer is:  ['[]', '[]']
******************** Question:
Cookie policy: []
Table: [] 
Answer:
[] 
Response:
[]
Len of comp:  1
Answer:  []
---> No.0  []
Len of comp:  1
Answer:  []
---> No.0  []
Answer is:  ['[]', '[]']
******************** Question:
Cookie policy: []
Table: [] 
Answer:
[] 
Response:
I've examined the provided cookie policy and nothing is mentioned about specific cookies. Since the policy is empty, the accurate response is an empty array.
Len of comp:  1
Answer:  []
---> No.0  I've examined the provided cookie policy and nothing is mentioned about specific cookies. Since the policy is empty, the accurate res

KeyboardInterrupt: 

In [ ]:
wandb.finish()
model.config.use_cache = True

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [ ]:
def infer():
  prompt = input("Enter your cookie policy")
  text = tokenizer.apply_chat_template([
      {'role': 'system', 'content': SYSTEM_PROMPT},
      {"role" : "user", "content" : prompt},
  ], tokenize = False, add_generation_prompt = True)

  from vllm import SamplingParams
  sampling_params = SamplingParams(
      temperature = 0.8,
      top_p = 0.95,
      max_tokens = 1024,
  )
  output = model.fast_generate(
      [text],
      sampling_params = sampling_params,
      lora_request = None,
  )[0].outputs[0].text

  print(output)

infer()

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [ ]:
model.save_lora("grpo_saved_lora")

Now we load the LoRA and test:

In [ ]:
text = tokenizer.apply_chat_template([
    {"role" : "system", "content" : SYSTEM_PROMPT},
    {"role" : "user", "content" : "Calculate pi."},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
